docs/src/adr/011adr-smt-arrays.md

```diff
@@ -169,11 +169,112 @@ The following changes will be made to implement the new encoding of sets:
+TLA+ set. The second array will store the mappings, having sort `<S1,...,Sn>` as its domain, with
+`Si` being the sort of argument `i`, and the sort of the function's codomain as its range. 
+The sorts of the array domain and range can be infinite, but the domain of the function itself,
+and by implication the number of mappings, will always be finite.
+
+To encode the TLA+ function `finSucc = [x \in {1,2,3} |-> x + 1 ]`, which computes the successors
+of integers from `1` to `3`, we first have to declare its domain, as shown below; tuples are
+represented here as per the OOPSLA'19 encoding.
+
+```
+(declare-sort Tuple_Int 0) ; Sort of <Int>
+(declare-const tuple_with_1 Tuple_Int) ; <1>
+(declare-const tuple_with_2 Tuple_Int) ; <2>
+(declare-const tuple_with_3 Tuple_Int) ; <3>
+
+(declare-const finSucc_domain_0 (Array Tuple_Int Bool))
+(declare-const finSucc_domain_1 (Array Tuple_Int Bool))
+(declare-const finSucc_domain_2 (Array Tuple_Int Bool))
+(declare-const finSucc_domain_3 (Array Tuple_Int Bool))
+
+(assert (= finSucc_domain_0 ((as const (Array Tuple_Int Bool)) false)))  ; {}
+(assert (= finSucc_domain_1 (store finSucc_domain_0 tuple_with_1 true))) ; {<1>}
+(assert (= finSucc_domain_2 (store finSucc_domain_1 tuple_with_2 true))) ; {<1>,<2>}
+(assert (= finSucc_domain_3 (store finSucc_domain_2 tuple_with_3 true))) ; {<1>,<2>,<3>}
+```
+
+The array storing the function's domain is used to guard the definition of the array storing the
+function's mappings, since mappings should only be present for values in the domain. The array
+storing the mappings of `finSucc` is shown below.
+
+```
+(declare-const finSucc_0 (Array Tuple_Int Int))
+(declare-const finSucc_1 (Array Tuple_Int Int))
+(declare-const finSucc_2 (Array Tuple_Int Int))
+(declare-const finSucc_3 (Array Tuple_Int Int))
+
+(assert (ite (select finSucc_domain_3 tuple_with_1)
+             (= finSucc_1 (store finSucc_0 tuple_with_1 2))
+             (= finSucc_1 finSucc_0)))
+(assert (ite (select finSucc_domain_3 tuple_with_2)
+             (= finSucc_2 (store finSucc_1 tuple_with_2 3))
+             (= finSucc_2 finSucc_1)))
+(assert (ite (select finSucc_domain_3 tuple_with_3)
+             (= finSucc_3 (store finSucc_2 tuple_with_3 4))
+             (= finSucc_3 finSucc_2)))
+```
+
+Note that, unlike with the new encoding for sets, we do not use constant arrays. The reason is that
+the function's domain cannot be altered, so the array has to constrain only the values in said domain.
+Function application can be done by simply accessing the array at the index of the passed arguments.
+A function application with arguments outside the function's domain leads to an unspecified result in
+TLA+, which is perfectly captured by unconstrained entries in the SMT array. Below are some examples
+of function application.
+
+```
+(assert (= (select finSucc_3 tuple_with_1) 2)) ; SAT
+(assert (= (select finSucc_3 tuple_with_2) 3)) ; SAT
+(assert (= (select finSucc_3 tuple_with_3) 4)) ; SAT
```

Comment:
It is worth adding a note that this is in sharp contrast with the OOPSLA encoding, which has to produce a number of constraints that is linear in the size of the set approximation.